Added docs for app health checks

ItalyPaleAle opened the pull request:
Fixes #2734

Also removes ServiceInvocation.NoDefaultContentType flag (fixes #2869)

Note: CLI flags listed here depend on https://github.com/dapr/cli/pull/1083

ItalyPaleAle committed:
Added docs for app health checks

Fixes #2734

Also removes ServiceInvocation.NoDefaultContentType flag (fixes #2869)

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Merge branch 'v1.9' into apphealth

msfussell commented:
For this need to add
1) Link to the community call video from call #70
2) Add or update a diagram to show how this works (optional for this PR, needed otherwise)

hhunter-ms commented on daprdocs/content/en/concepts/observability-concept.md:
```diff
@@ -7,38 +7,48 @@ description: >
 ---
 
-When building an application, understanding how the system is behaving is an important part of operating it - this includes having the ability to observe the internal calls of an application, gauging its performance and becoming aware of problems as soon as they occur. This is challenging for any system, but even more so for a distributed system comprised of multiple microservices where a flow, m ...
+When building an application, understanding how the system is behaving is an important part of operating it–this includes having the ability to observe the internal calls of an application, gauging its performance and becoming aware of problems as soon as they occur. This is challenging for any system, but even more so for a distributed system comprised of multiple microservices where a flow, mad ...
```
```suggestion
When building and operating an application, it's important to understand how the system is behaving. You need to be able to:
- Observe the internal calls of an application.
- Gauge the application's performance.
- Become aware of problems as soon as they occur. 

This is challenging for any system, especially a distributed system comprised of multiple microservices where a flow, made of several calls, may start in one microservice and continue in another. Observability is critical in production environments, and proves useful during development to:
- Understand bottlenecks
- Improve performance
- Perform basic debugging across the span of microservices
```

hhunter-ms commented on daprdocs/content/en/concepts/observability-concept.md:
```diff
@@ -7,38 +7,48 @@ description: >
+When building an application, understanding how the system is behaving is an important part of operating it–this includes having the ability to observe the internal calls of an application, gauging its performance and becoming aware of problems as soon as they occur. This is challenging for any system, but even more so for a distributed system comprised of multiple microservices where a flow, mad ...
 
-While some data points about an application can be gathered from the underlying infrastructure (e.g. memory consumption, CPU usage), other meaningful information must be collected from an "application-aware" layer - one that can show how an important series of calls is executed across microservices. This usually means a developer must add some code to instrument an application for this purpose. O ...
+While some data points about an application can be gathered from the underlying infrastructure (e.g. memory consumption, CPU usage), other meaningful information must be collected from an "application-aware" layer: one that can show how an important series of calls is executed across microservices. This usually means a developer must add some code to instrument an application for this purpose. Of ...
```
```suggestion
While some data points about your application can be gathered from the underlying infrastructure (memory consumption, CPU usage, etc.), other meaningful information must be collected from an "application-aware" layer. This layer can show how an important series of calls is executed across microservices. Typically, you implement this layer by adding some code to instrument an application for this purpose. Instrumentation code is not part of the core logic of the application, but is simply meant to send collected data (traces and metrics) to an external monitoring tool or service that stores, visualizes, and analyzes the information.
```

hhunter-ms commented on daprdocs/content/en/concepts/observability-concept.md:
```diff
@@ -7,38 +7,48 @@ description: >
+While some data points about an application can be gathered from the underlying infrastructure (e.g. memory consumption, CPU usage), other meaningful information must be collected from an "application-aware" layer: one that can show how an important series of calls is executed across microservices. This usually means a developer must add some code to instrument an application for this purpose. Of ...
 
-Having to maintain this code, which is not part of the core logic of the application, is another burden on the developer, sometimes requiring understanding the monitoring tools' APIs, using additional SDKs etc. This instrumentation may also add to the portability challenges of an application, which may require different instrumentation depending on where the application is deployed. For example,  ...
+Having to maintain this code, which is not part of the core logic of the application, is another burden on the developer, sometimes requiring understanding the monitoring tools' APIs, using additional SDKs, etc. This may also add to the portability challenges of an application, requiring different instrumentation depending on where the application is deployed. For example, different cloud provide ...
```
```suggestion
Maintaining instrumentation code can be cumbersome, requiring you to understand the monitoring tools' APIs, use additional SDKs, etc. It can also decrease your application's portability, requiring different instrumentation depending on where the application is deployed. For example, different cloud providers offer different monitoring solutions and an on-premises deployment might require a self-hosted solution.
```

hhunter-ms commented on daprdocs/content/en/concepts/observability-concept.md:
```diff
@@ -7,38 +7,48 @@ description: >
 ## Observability for your application with Dapr
-When building an application which leverages Dapr building blocks to perform service-to-service calls and pub/sub messaging, Dapr offers an advantage with respect to [distributed tracing]({{<ref tracing>}}). Because this inter-service communication flows through the Dapr sidecar, the sidecar is in a unique position to offload the burden of application-level instrumentation.
+
+When building an application which leverages Dapr building blocks to perform service-to-service calls and pub/sub messaging, Dapr offers an advantage with respect to [distributed tracing]({{<ref tracing>}}). Because this inter-service communication flows through the Dapr sidecar, Dapr is in a unique position to offload the burden of application-level instrumentation.
```
```suggestion
Dapr offers an advantage to applications that leverage its service-to-service invocation and pub/sub building blocks in respect to [distributed tracing]({{<ref tracing>}}). Because this inter-service communication flows through the Dapr sidecar, Dapr is in a unique position to offload the burden of application-level instrumentation.
```

hhunter-ms commented on daprdocs/content/en/concepts/observability-concept.md:
```diff
@@ -7,38 +7,48 @@ description: >
+
 The Dapr sidecar exposes an HTTP endpoint for [health checks]({{<ref sidecar-health.md>}}). With this API, user code or hosting environments can probe the Dapr sidecar to determine its status and identify issues with sidecar readiness.
+
+Conversely, Dapr can be configured to probe for the [health of your application]({{ <ref app-health.md> }}), and react to changes in the app's health including by stopping pub/sub subscriptions and short-circuiting service invocation calls.
```
```suggestion
Conversely, Dapr can be configured to probe for the [health of your application]({{ <ref app-health.md> }}), and react to changes in the app's health, including stopping pub/sub subscriptions and short-circuiting service invocation calls.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+description: Reacting to apps' health status changes
+---
+
+App health checks is a feature of Dapr that allows probing for the health of your application and reacting to status changes.
```
```suggestion
App health checks is a preview feature that allows probing for the health of your application and reacting to status changes.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+App health checks is a feature of Dapr that allows probing for the health of your application and reacting to status changes.
+
+Applications can become unresponsive for a variety of reasons: for example, they could be too busy to accept new work, or could be crashed or in a deadlock state. Sometimes the condition can be transitory, for example if the app is just busy (and will eventually be able to resume accepting new work), or if the application is being restarted for whatever reason and is in its initialization phase.
```
```suggestion
Applications can become unresponsive for a variety of reasons, such as:

 - Too busy to accept new work
 - Crashed state
 - In a deadlock state
 - In initialization phase after being restarted
```

ItalyPaleAle replied:
I would prefer this as inline as it was just a list of examples, not exhaustive and also more informational than other

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+When app health checks are enabled, the Dapr *runtime* (also known as the *sidecar*) will periodically poll your application via HTTP or gRPC calls.
+
+When it detects a failure in the app's health, Dapr stops accepting new work on behalf of the application:
```
```suggestion
When it detects a failure in the app's health, Dapr stops accepting new work on behalf of the application by:
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+When it detects a failure in the app's health, Dapr stops accepting new work on behalf of the application:
+
+- Unsubscribes from all pub/sub subscriptions
```
```suggestion
- Unsubscribing from all pub/sub subscriptions
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+When it detects a failure in the app's health, Dapr stops accepting new work on behalf of the application:
+
+- Unsubscribes from all pub/sub subscriptions
+- Stops all input bindings
```
```suggestion
- Stopping all input bindings
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+- Unsubscribes from all pub/sub subscriptions
+- Stops all input bindings
+- Short-circuits all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application
```
```suggestion
- Short-circuiting all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+- Stops all input bindings
+- Short-circuits all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application
+
+These changes are meant to be temporary, and Dapr will resume normal operations once it detects that the application becomes responsive again.
```
```suggestion
These changes are meant to be temporary, and Dapr will resume normal operations once it detects that the application is responsive again.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+App health checks are disabled by default.
+
+> App health checks in Dapr are meant to be complementary to, and not replace, any platform-level health checks, for example [liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) when running on Kubernetes.  
```
```suggestion
App health checks in Dapr are meant to be complementary to, and not replace, any platform-level health checks, like [liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/), when running on Kubernetes.  
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+App health checks are disabled by default.
+
+> App health checks in Dapr are meant to be complementary to, and not replace, any platform-level health checks, for example [liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) when running on Kubernetes.  
+> Platform-level health checks (or liveness probes) are generally meant to ensure that the application is running, and cause the platform to restart the application (or, in case of Kubernetes, pod) in case of failures.  
```
```suggestion
Platform-level health checks (or liveness probes) generally ensure that the application is running, and cause the platform to restart the application (or, in case of Kubernetes, pod) in case of failures.  
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+These changes are meant to be temporary, and Dapr will resume normal operations once it detects that the application becomes responsive again.
+
+App health checks are disabled by default.
+
```
```suggestion

## App health checks vs platform-level health checks

```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+> App health checks in Dapr are meant to be complementary to, and not replace, any platform-level health checks, for example [liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) when running on Kubernetes.  
+> Platform-level health checks (or liveness probes) are generally meant to ensure that the application is running, and cause the platform to restart the application (or, in case of Kubernetes, pod) in case of failures.  
+> Dapr's app health checks are instead concerned with not bringing more work to an application that is not currently in a condition to accept it, but that it is expected to be able to resume doing so *eventually*. Goals include not bringing more load to an application that is already overloaded, as well as being "polite" and not taking messages from the likes of queues or pub/subs when Dapr knows ...
```
```suggestion
Unlike platform-level health checks, Dapr's app health checks focus on pausing work to an application that is currently unable to accept it, but is expected to be able to resume accepting work *eventually*. Goals include:
- Not bringing more load to an application that is already overloaded.
- "Politely" not taking messages from queues or pub/subs when Dapr knows the application won't be able to process them. 

In this regard, Dapr's app health checks are "softer", waiting for an application to be able to process work, rather than terminating the running process in a "hard" way.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+## Configuring app health checks
+
+{{% alert title="Note" color="primary" %}}
+App health checks are currently a **preview feature** and require the `AppHealthCheck` feature flag to be enabled. Please refer to the documentation for [enabling preview features]({{<ref support-preview-features>}}).
```
```suggestion
App health checks are currently a **preview feature** and require the `AppHealthCheck` feature flag to be enabled. Refer to the documentation for [enabling preview features]({{<ref support-preview-features>}}).
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+App health checks are currently a **preview feature** and require the `AppHealthCheck` feature flag to be enabled. Please refer to the documentation for [enabling preview features]({{<ref support-preview-features>}}).
+{{% /alert %}}
+
+App health checks are enabled with the `--enable-app-health-check` CLI flag, or the `dapr.io/enable-app-health-check: true` annotation when running on Kubernetes; they are disabled by default. Adding this flag is both necessary and sufficient to enable app health checks with the default options.
```
```suggestion
App health checks are disabled by default, but can be enabled with either:

- The `--enable-app-health-check` CLI flag; or
- The `dapr.io/enable-app-health-check: true` annotation when running on Kubernetes.

Adding this flag is both necessary and sufficient to enable app health checks with the default options.
```

ItalyPaleAle replied:
I think the list here doesn't reflect that it's an `or`, sounds more like an `and`

hhunter-ms replied:
I agree. Updated my suggestion... is that better?

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+These changes are meant to be temporary, and Dapr will resume normal operations once it detects that the application becomes responsive again.
+
+App health checks are disabled by default.
```
```suggestion
```

Ok to remove since we mention this later?

ItalyPaleAle replied:
Your call, but I do believe  _repetita iuvant_ sometimes :)

hhunter-ms replied:
ok! I'm open either way :)

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+| `--app-health-probe-timeout`  | `dapr.io/app-health-probe-timeout`  | Timeout in *milliseconds* for health probe requests | `500` |
+| `--app-health-threshold`      | `dapr.io/app-health-threshold"`     | Max number of consecutive failures before the app is considered un-healthy | `3` |
+
+> See the [full reference]({{<ref arguments-annotations-overview>}}) for Dapr arguments and annotations for all options and how to enable them.
```
```suggestion
> See the [full Dapr arguments and annotations reference]({{<ref arguments-annotations-overview>}}) for all options and how to enable them.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+> See the [full reference]({{<ref arguments-annotations-overview>}}) for Dapr arguments and annotations for all options and how to enable them.
+
+Additionally, app health checks are impacted by the protocol used for the app channel, which is configured with the `--app-protocol` flag or the `dapr.io/app-protocol` annotation; supported values are `http` (default) or `grpc`.
```
```suggestion
Additionally, app health checks are impacted by the protocol used for the app channel, which is configured with the `--app-protocol` flag (self-hosted) or the `dapr.io/app-protocol` annotation (Kubernetes); supported values are `http` (default) or `grpc`.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+### Health check paths
+
+When using HTTP for `app-protocol`, Dapr will perform health probes by making a HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
```
```suggestion
When using HTTP for `app-protocol`, Dapr will perform health probes by making an HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
```

ItalyPaleAle replied:
Isn't it "an HTTP" because HTTP is pronounced "aitch"?

hhunter-ms replied:
Yep, even though HTTP technically starts with a consonant, it sounds like it starts with a vowel... it's all in the sound :)

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+### Health check paths
+
+When using HTTP for `app-protocol`, Dapr will perform health probes by making a HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
+For your app to be considered healthy, the response must have an HTTP status code in the 200-299 range; any other status code is considered a failure. Dapr is only concerned with the status code of the response, and will ignore any response header or body.
```
```suggestion
For your app to be considered healthy, the response must have an HTTP status code in the 200-299 range. Any other status code is considered a failure. Dapr is only concerned with the status code of the response, and will ignore any response header or body.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+When using gRPC for the app channel, Dapr will invoke the method `/dapr.proto.runtime.v1.AppCallbackHealthCheck/HealthCheck` in your application. Most likely, you will use a Dapr SDK to implement the handler for this method.
+
+While responding to a health probe request, your app *may* decide to perform additional internal health checks to determine if it's actually in a state where it's ready to process work from the Dapr runtime. However, this is not required, and it's a choice that depends on your application's needs.
```
```suggestion
While responding to a health probe request, your app *may* decide to perform additional internal health checks to determine if it's ready to process work from the Dapr runtime. However, this is not required; it's a choice that depends on your application's needs.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+### Intervals, timeouts, and thresholds
+
+When app health checks are enabled, by default Dapr probes your application every 5 seconds; you can configure the interval, in seconds, with `app-health-probe-interval`. These probes happen regularly, regardless of whether your application is healthy or not.
```
```suggestion
When app health checks are enabled, by default Dapr probes your application every 5 seconds. You can configure the interval, in seconds, with `app-health-probe-interval`. These probes happen regularly, regardless of whether your application is healthy or not.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+When app health checks are enabled, by default Dapr probes your application every 5 seconds; you can configure the interval, in seconds, with `app-health-probe-interval`. These probes happen regularly, regardless of whether your application is healthy or not.
+
+When the Dapr runtime (sidecar) is first started, Dapr waits for a first, successful health probe before considering the app healthy. This means that pub/sub subscriptions, input bindings, and service invocation won't be enabled for your app until a first health check is complete and successful.
```
```suggestion
When the Dapr runtime (sidecar) is initially started, Dapr waits for a successful health probe before considering the app healthy. This means that pub/sub subscriptions, input bindings, and service invocation won't be enabled for your app until this first health check is complete and successful.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+
+Health probe requests are considered successful if the application sends a successful response (as explained above) within the timeout configured in `app-health-probe-timeout`. The default value is 500, corresponding to 500 milliseconds (i.e. half a second).
+
+Before Dapr considers an app to have entered an un-healthy state, it will wait for `app-health-threshold` consecutive failures, whose default value is 3.  
```
```suggestion
Before Dapr considers an app to have entered an unhealthy state, it will wait for `app-health-threshold` consecutive failures, whose default value is 3.  
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+Health probe requests are considered successful if the application sends a successful response (as explained above) within the timeout configured in `app-health-probe-timeout`. The default value is 500, corresponding to 500 milliseconds (i.e. half a second).
+
+Before Dapr considers an app to have entered an un-healthy state, it will wait for `app-health-threshold` consecutive failures, whose default value is 3.  
+With the default value, it means that your application must be fail health probes 3 times *in a row* to be considered un-healthy.  
```
```suggestion
This default value means that your application must fail health probes 3 times *in a row* to be considered unhealthy.  
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+Before Dapr considers an app to have entered an un-healthy state, it will wait for `app-health-threshold` consecutive failures, whose default value is 3.  
+With the default value, it means that your application must be fail health probes 3 times *in a row* to be considered un-healthy.  
+If you set the threshold to 1, any failure will cause Dapr to assume your app is unhealthy and will stop delivering work to it.  
+A threshold greater than 1 can help exclude transient failures due to external circumstances; the right value for your application depends on your requirements.
```
```suggestion
A threshold greater than 1 can help exclude transient failures due to external circumstances. The right value for your application depends on your requirements.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+If you set the threshold to 1, any failure will cause Dapr to assume your app is unhealthy and will stop delivering work to it.  
+A threshold greater than 1 can help exclude transient failures due to external circumstances; the right value for your application depends on your requirements.
+
+Note that thresholds only apply to failures. A single successful response is enough for Dapr to consider your app to be healthy and resume normal operations.
```
```suggestion
Thresholds only apply to failures. A single successful response is enough for Dapr to consider your app to be healthy and resume normal operations.
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/sidecar-health.md:
```diff
@@ -2,63 +2,63 @@
 ---
 
-Dapr provides a way to [determine its health using an HTTP `/healthz` endpoint]({{< ref health_api.md >}}). With this endpoint, the Dapr process, or sidecar, can be:
+Dapr provides a way to[determine its health using an [HTTP `/healthz` endpoint]({{< ref health_api.md >}}). With this endpoint, the *daprd* process, or sidecar, can be:
```
```suggestion
Dapr provides a way to [determine its health using an [HTTP `/healthz` endpoint]({{< ref health_api.md >}}). With this endpoint, the *daprd* process, or sidecar, can be:
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/sidecar-health.md:
```diff
@@ -67,20 +67,21 @@ readinessProbe:
 As mentioned above, this configuration is done automatically by the Sidecar Injector service. This section describes the specific values that are set on the liveness and readiness probes.
 
-Dapr has its HTTP health endpoint `/v1.0/healthz` on port 3500, This can be used with Kubernetes for readiness and liveness probe. When the Dapr sidecar is injected , the readiness and liveness probes are configured in the pod configuration file with the following values.
+Dapr has its HTTP health endpoint `/v1.0/healthz` on port 3500, This can be used with Kubernetes for readiness and liveness probe. When the Dapr sidecar is injected, the readiness and liveness probes are configured in the pod configuration file with the following values:
```
```suggestion
Dapr has its HTTP health endpoint `/v1.0/healthz` on port 3500. This can be used with Kubernetes for readiness and liveness probe. When the Dapr sidecar is injected, the readiness and liveness probes are configured in the pod configuration file with the following values:
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/sidecar-health.md:
````diff
@@ -2,63 +2,63 @@
 ```
 
-In the above *example*, the `periodSeconds` field specifies that the kubelet should perform a liveness probe every 3 seconds. The `initialDelaySeconds` field tells the kubelet that it should wait 3 seconds before performing the first probe. To perform a probe, the kubelet sends an HTTP GET request to the server that is running in the container and listening on port 8080 in this example. If the ha ...
+In the above example, the `periodSeconds` field specifies that the kubelet should perform a liveness probe every 3 seconds. The `initialDelaySeconds` field tells the kubelet that it should wait 3 seconds before performing the first probe. To perform a probe, the kubelet sends an HTTP GET request to the server that is running in the container and listening on port 8080 in this example. If the hand ...
````
```suggestion
In the above example, the `periodSeconds` field specifies that the kubelet should perform a liveness probe every 3 seconds. The `initialDelaySeconds` field tells the kubelet that it should wait 3 seconds before performing the first probe. To perform a probe, the kubelet sends an HTTP GET request to the server that is running in the container and listening on port 8080. If the handler for the server's `/healthz` path returns a success code, the kubelet considers the container to be alive and healthy. If the handler returns a failure code, the kubelet kills the container and restarts it.
```

hhunter-ms commented on daprdocs/content/en/reference/arguments-annotations-overview.md:
```diff
@@ -8,7 +8,7 @@ aliases:
 ---
 
-This table is meant to help users understand the equivalent options for running Dapr sidecars in different contexts--via the [CLI]({{< ref cli-overview.md >}}) directly, via daprd, or on [Kubernetes]({{< ref kubernetes-overview.md >}}) via annotations.
+This table is meant to help users understand the equivalent options for running Dapr sidecars in different contexts–via the [CLI]({{< ref cli-overview.md >}}) directly, via daprd, or on [Kubernetes]({{< ref kubernetes-overview.md >}}) via annotations.
```
```suggestion
This table is meant to help users understand the equivalent options for running Dapr sidecars in different contexts, via either:

 - The [CLI]({{< ref cli-overview.md >}}) directly;
 - _daprd_; or
 - [Kubernetes]({{< ref kubernetes-overview.md >}}) annotations.
```

ItalyPaleAle replied:
Here too, I think a list gives less of a sense that it's an "or". Also, it looks very un-balanced visually

hhunter-ms replied:
Ok! I like lists if you can't tell haha. I'm fine either way though :)

hhunter-ms commented on daprdocs/content/en/reference/arguments-annotations-overview.md:
```diff
@@ -38,6 +38,11 @@ This table is meant to help users understand the equivalent options for running
+| `--app-health-check-path` | `--app-health-check-path` | | `dapr.io/app-health-check-path` | Path that Dapr invokes for health probes when the app channel is HTTP (this value is ignored if the app channel is using gRPC). Requires app health checks to be enabled. Default is `/health` |
+| `--app-health-probe-interval` | `--app-health-probe-interval` | | `dapr.io/app-health-probe-interval` | Number of *seconds* between each health probe. Requires app health checks to be enabled. Default is `5` |
+| `--app-health-probe-timeout` | `--app-health-probe-timeout` | | `dapr.io/app-health-probe-timeout` | Timeout in *milliseconds* for health probe requests. Requires app health checks to be enabled. Default is `500` |
+| `--app-health-threshold` | `--app-health-threshold` | | `dapr.io/app-health-threshold"` | Max number of consecutive failures before the app is considered un-healthy. Requires app health checks to be enabled. Default is `3` |
```
```suggestion
| `--app-health-threshold` | `--app-health-threshold` | | `dapr.io/app-health-threshold"` | Max number of consecutive failures before the app is considered unhealthy. Requires app health checks to be enabled. Default is `3` |
```

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,73 @@
+| `--app-health-check-path`     | `dapr.io/app-health-check-path`     | Path that Dapr invokes for health probes when the app channel is HTTP (this value is ignored if the app channel is using gRPC) | `/health` |
+| `--app-health-probe-interval` | `dapr.io/app-health-probe-interval` | Number of *seconds* between each health probe | `5` |
+| `--app-health-probe-timeout`  | `dapr.io/app-health-probe-timeout`  | Timeout in *milliseconds* for health probe requests | `500` |
+| `--app-health-threshold`      | `dapr.io/app-health-threshold"`     | Max number of consecutive failures before the app is considered un-healthy | `3` |
```
```suggestion
| `--app-health-threshold`      | `dapr.io/app-health-threshold"`     | Max number of consecutive failures before the app is considered unhealthy | `3` |
```

hhunter-ms reviewed (changes requested):
quick grammar/readability

msfussell committed:
Merge branch 'v1.9' into apphealth

ItalyPaleAle committed:
Apply suggestions from code review

Co-authored-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
tweaks

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Update daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md

Co-authored-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Update daprdocs/content/en/concepts/observability-concept.md

Co-authored-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Update daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md

Co-authored-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Added example

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Merge branch 'v1.9' into apphealth

msfussell committed:
Merge branch 'v1.9' into apphealth

Signed-off-by: Mark Fussell <markfussell@gmail.com>

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+description: Reacting to apps' health status changes
+---
+
+App health checks is a preview feature that allows probing for the health of your application and reacting to status changes.
```
```suggestion
App health checks is a feature that allows probing for the health of your application and reacting to status changes.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+App health checks is a preview feature that allows probing for the health of your application and reacting to status changes.
+
+Applications can become unresponsive for a variety of reasons: for example, they could be too busy to accept new work, or could be crashed or in a deadlock state. Sometimes the condition can be transitory, for example if the app is just busy (and will eventually be able to resume accepting new work), or if the application is being restarted for whatever reason and is in its initialization phase.
```
```suggestion
Applications can become unresponsive for a variety of reasons: for example, they could be too busy to accept new work, could have crashed, or be in a deadlock state. Sometimes the condition can be transitory, for example if the app is just busy (and will eventually be able to resume accepting new work), or if the application is being restarted for whatever reason and is in its initialization phase.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+Applications can become unresponsive for a variety of reasons: for example, they could be too busy to accept new work, or could be crashed or in a deadlock state. Sometimes the condition can be transitory, for example if the app is just busy (and will eventually be able to resume accepting new work), or if the application is being restarted for whatever reason and is in its initialization phase.
+
+When app health checks are enabled, the Dapr *runtime* (also known as the *sidecar*) will periodically poll your application via HTTP or gRPC calls.
```
```suggestion
When app health checks are enabled, the Dapr *runtime* (sidecar) periodically polls your application via HTTP or gRPC calls.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+When it detects a failure in the app's health, Dapr stops accepting new work on behalf of the application by:
+
+- Unsubscribing from all pub/sub subscriptions
+- Stopping all input bindings
```
```suggestion
- Stopping all incoming requests from input bindings
```

ItalyPaleAle replied:
Actually this is not correct. It stops the input bindings entirely doesn't just stop sending incoming requests from the app

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+- Unsubscribing from all pub/sub subscriptions
+- Stopping all input bindings
+- Short-circuiting all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application
```
```suggestion
- Stopping all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application.
```

ItalyPaleAle replied:
I think short-circuiting here gives a better idea of the fact that the requests can still happen but the sidecar terminates them earlier, before even trying to hit the app.

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+When it detects a failure in the app's health, Dapr stops accepting new work on behalf of the application by:
+
+- Unsubscribing from all pub/sub subscriptions
```
```suggestion
- Unsubscribing from all pub/sub subscription messages
```

ItalyPaleAle replied:
Same, this actually stop the subscriptions and not just the messages

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+- Stopping all input bindings
+- Short-circuiting all service-invocation requests, which terminate in the Dapr runtime and are not forwarded to the application
+
+These changes are meant to be temporary, and Dapr will resume normal operations once it detects that the application is responsive again.
```
```suggestion
These changes are meant to be temporary, and Dapr resumes normal operations once it detects that the application is responsive again.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+Unlike platform-level health checks, Dapr's app health checks focus on pausing work to an application that is currently unable to accept it, but is expected to be able to resume accepting work *eventually*. Goals include:
+
+- Not bringing more load to an application that is already overloaded.
+- Do the "polite" thing by not taking messages from queues or pub/subs when Dapr knows the application won't be able to process them.
```
```suggestion
- Do the "polite" thing by not taking messages from queues, bindings, or pub/sub brokers when Dapr knows the application won't be able to process them.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+### Health check paths
+
+When using HTTP for `app-protocol`, Dapr will perform health probes by making an HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
```
```suggestion
When using HTTP for `app-protocol`, Dapr performs health probes by making an HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+### Health check paths
+
+When using HTTP for `app-protocol`, Dapr will perform health probes by making an HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
+For your app to be considered healthy, the response must have an HTTP status code in the 200-299 range. Any other status code is considered a failure. Dapr is only concerned with the status code of the response, and will ignore any response header or body.
```
```suggestion
For your app to be considered healthy, the response must have an HTTP status code in the 200-299 range. Any other status code is considered a failure. Dapr is only concerned with the status code of the response, and ignores any response header or body.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+When using HTTP for `app-protocol`, Dapr will perform health probes by making an HTTP call to the path specified in `app-health-check-path`, which is `/health` by default.  
+For your app to be considered healthy, the response must have an HTTP status code in the 200-299 range. Any other status code is considered a failure. Dapr is only concerned with the status code of the response, and will ignore any response header or body.
+
+When using gRPC for the app channel, Dapr will invoke the method `/dapr.proto.runtime.v1.AppCallbackHealthCheck/HealthCheck` in your application. Most likely, you will use a Dapr SDK to implement the handler for this method.
```
```suggestion
When using gRPC for the app channel, Dapr invokes the method `/dapr.proto.runtime.v1.AppCallbackHealthCheck/HealthCheck` in your application. Most likely, you will use a Dapr SDK to implement the handler for this method.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+When app health checks are enabled, by default Dapr probes your application every 5 seconds. You can configure the interval, in seconds, with `app-health-probe-interval`. These probes happen regularly, regardless of whether your application is healthy or not.
+
+When the Dapr runtime (sidecar) is initially started, Dapr waits for a successful health probe before considering the app healthy. This means that pub/sub subscriptions, input bindings, and service invocation won't be enabled for your app until this first health check is complete and successful.
```
```suggestion
When the Dapr runtime (sidecar) is initially started, Dapr waits for a successful health probe before considering the app healthy. This means that pub/sub subscriptions, input bindings, and service invocation requests won't be enabled for your application until this first health check is complete and successful.
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+Health probe requests are considered successful if the application sends a successful response (as explained above) within the timeout configured in `app-health-probe-timeout`. The default value is 500, corresponding to 500 milliseconds (i.e. half a second).
+
+Before Dapr considers an app to have entered an unhealthy state, it will wait for `app-health-threshold` consecutive failures, whose default value is 3.  
```
```suggestion
Before Dapr considers an app to have entered an unhealthy state, it waits for `app-health-threshold` consecutive failures, whose default value is 3.  
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+Before Dapr considers an app to have entered an unhealthy state, it will wait for `app-health-threshold` consecutive failures, whose default value is 3.  
+This default value means that your application must fail health probes 3 times *in a row* to be considered unhealthy.  
+If you set the threshold to 1, any failure will cause Dapr to assume your app is unhealthy and will stop delivering work to it.  
```
```suggestion
If you set the threshold to 1, any failure causes Dapr to assume your app is unhealthy and will stop delivering work to it.  
```

msfussell commented on daprdocs/content/en/developing-applications/building-blocks/observability/sidecar-health.md:
```diff
@@ -2,63 +2,63 @@
+- Determined for readiness and liveness
 
-The Dapr `/healthz` endpoint can be used by health probes from the application hosting platform. This topic describes how Dapr integrates with probes from different hosting platforms.
+The Dapr `/healthz` endpoint can be used by health probes from the application hosting platform (e.g. Kubernetes). This topic describes how Dapr integrates with probes from different hosting platforms.
```
```suggestion
The Dapr `/healthz` endpoint can be used by health probes from the application hosting platform (for example Kubernetes). This topic describes how Dapr integrates with probes from different hosting platforms.
```

ItalyPaleAle committed:
Update daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md

Co-authored-by: Mark Fussell <markfussell@gmail.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

msfussell commented:
@ItalyPaleAle @hhunter-ms  

Feedback
1) You also need to update the Dapr `run` CLI command here https://docs.dapr.io/reference/cli/dapr-run/
2) Can you include a link to the Community call video. Here is the text to embed into the app-health.md

Watch [this video for an overview of using the application health checks](https://youtu.be/srczBuOsAkI?t=533):

<iframe width="360" height="315" src="https://www.youtube-nocookie.com/embed/srczBuOsAkI?start=533" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

3) @hhunter-ms - Can you help drawn a diagram for this article. I suggest having a picture of the Dapr sidecar calling onto a /health HTTP endpoint on the app. @ItalyPaleAle - Could you advise here?

hhunter-ms commented on daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md:
```diff
@@ -0,0 +1,144 @@
+
+{{% /codetab %}}
+
+{{< /tabs >}}
```
```suggestion
{{< /tabs >}}

## Demo

Watch [this video for an overview of using the application health checks](https://youtu.be/srczBuOsAkI?t=533):

<iframe width="360" height="315" src="https://www.youtube-nocookie.com/embed/srczBuOsAkI?start=533" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

ItalyPaleAle committed:
Update daprdocs/content/en/developing-applications/building-blocks/observability/app-health.md

Co-authored-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

hhunter-ms commented:
@ItalyPaleAle @msfussell how is this?
![image](https://user-images.githubusercontent.com/94493363/195395295-3e5cd52e-0125-44d8-8f65-146bca61080e.png)

ItalyPaleAle committed:
Merge branch 'v1.9' of https://github.com/dapr/docs into apphealth

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Merge branch 'v1.9' of https://github.com/dapr/docs into apphealth

ItalyPaleAle commented:
@hhunter-ms thanks for making that :)

I think it looks good but if the flag `--app-health-check-path` is `/healthz`, then the endpoint on the right should be `/healthz` too. Otherwise, i think it's good!

(Worth also adding `--app-protocol http` to make it explicit?)

ItalyPaleAle committed:
Updated `dapr run` CLI ref

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
Added image for app health

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle committed:
fixes

Signed-off-by: ItalyPaleAle <43508+ItalyPaleAle@users.noreply.github.com>

ItalyPaleAle commented:
@msfussell Thanks to @hhunter-ms's help:

1. Added the image
2. Added video
3. Updated the "dapr run" CLI ref page
4. Fixed all merge conflicts

Ready to go?

hhunter-ms committed:
Merge branch 'v1.9' into apphealth

Signed-off-by: Hannah Hunter <94493363+hhunter-ms@users.noreply.github.com>

msfussell reviewed (commented):
I missed this review post

ItalyPaleAle committed:
Apply suggestions from code review

Co-authored-by: Mark Fussell <markfussell@gmail.com>
Signed-off-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>
